src/hotspot/share/runtime/thread.cpp

```diff
@@ -3593,6 +3594,13 @@ void Threads::remove(JavaThread* p, bool is_daemon) {
   // that we do not remove thread without safepoint code notice
   { MonitorLocker ml(Threads_lock);
 
+    if (ThreadIdTable::is_initialized()) {
+      // This cleanup must be done before the current thread's GC barrier
+      // is detached since we need to touch the threadObj oop.
+      jlong tid = SharedRuntime::get_java_tid(p);
+      ThreadIdTable::remove_thread(tid);
+    }
+
```

Comment:
I don't like exposing `ThreadSMR` internal details like this. Can we at least make this a little `ThreadSMRSupport` method that we call from here.

Or ... can we just move `ThreadsSMRSupport::remove_thread(p);` to before `on_thread_detach`?

Or can we grab the pid before `on_thread_detach` and pass it to `ThreadsSMRSupport::remove_thread`?

Reply:
Ummmm... The code that I moved is not a TheadSMR internal detail. That code
is part of the M&M support that was added by Daniil in:

JDK-8185005 Improve performance of ThreadMXBean.getThreadInfo(long ids[], int maxDepth)
https://bugs.openjdk.org/browse/JDK-8185005

I don't remember why Daniil put that code in ThreadsSMRSupport::remove_thread(p),
but it is just as appropriate to put it in Threads::remove(). That M&M cleanup code
is part of the cleanup necessary when a JavaThread is removed. The important part
is to call that code after the Threads_lock is grabbed.

Reply:
Thanks Dan, I had forgotten about that and just assumed it was ThreadSMR related. Moving it as you did is now not a concern.